saga/saga_test.go

```diff
@@ -0,0 +1,331 @@
+		log: sagaLogMock,
+	}
+	state, err := s.StartSaga("testSaga", nil)
+	state, err = s.StartTask(state, "task1")
+
+	state, err = s.EndTask(state, "task1", nil)
+	if err == nil {
+		t.Error(fmt.Sprintf("Expected EndTask to not return an error when write to SagaLog Fails"))
+	}
+}
+
+func TestStartCompTask(t *testing.T) {
+	entry := StartCompTaskMessageFactory("testSaga", "task1")
+
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	sagaLogMock := NewMockSagaLog(mockCtrl)
+	sagaLogMock.EXPECT().StartSaga("testSaga", nil)
+	sagaLogMock.EXPECT().LogMessage(StartTaskMessageFactory("testSaga", "task1"))
+	sagaLogMock.EXPECT().LogMessage(AbortSagaMessageFactory("testSaga"))
+	sagaLogMock.EXPECT().LogMessage(entry)
+
+	s := Saga{
+		log: sagaLogMock,
+	}
+	state, err := s.StartSaga("testSaga", nil)
+	state, err = s.StartTask(state, "task1")
+	state, err = s.AbortSaga(state)
+
+	state, err = s.StartCompensatingTask(state, "task1")
+	if err != nil {
+		t.Error(fmt.Sprintf("Expected StartCompensatingTask to not return an error"))
+	}
+}
+
+func TestStartCompTaskLogError(t *testing.T) {
+	entry := StartCompTaskMessageFactory("testSaga", "task1")
+
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	sagaLogMock := NewMockSagaLog(mockCtrl)
+	sagaLogMock.EXPECT().StartSaga("testSaga", nil)
+	sagaLogMock.EXPECT().LogMessage(StartTaskMessageFactory("testSaga", "task1"))
+	sagaLogMock.EXPECT().LogMessage(AbortSagaMessageFactory("testSaga"))
+	sagaLogMock.EXPECT().LogMessage(entry).Return(errors.New("Failed to Log StartCompTask Message"))
+
+	s := Saga{
+		log: sagaLogMock,
+	}
+	state, err := s.StartSaga("testSaga", nil)
+	state, err = s.StartTask(state, "task1")
+	state, err = s.AbortSaga(state)
+
+	state, err = s.StartCompensatingTask(state, "task1")
+	if err == nil {
+		t.Error(fmt.Sprintf("Expected StartCompTask to not return an error when write to SagaLog Fails"))
+	}
+}
+
+func TestEndCompTask(t *testing.T) {
+	entry := EndCompTaskMessageFactory("testSaga", "task1", nil)
+
+	mockCtrl := gomock.NewController(t)
+	defer mockCtrl.Finish()
+
+	sagaLogMock := NewMockSagaLog(mockCtrl)
+	sagaLogMock.EXPECT().StartSaga("testSaga", nil)
+	sagaLogMock.EXPECT().LogMessage(StartTaskMessageFactory("testSaga", "task1"))
+	sagaLogMock.EXPECT().LogMessage(AbortSagaMessageFactory("testSaga"))
+	sagaLogMock.EXPECT().LogMessage(StartCompTaskMessageFactory("testSaga", "task1"))
+	sagaLogMock.EXPECT().LogMessage(entry)
+
+	s := Saga{
+		log: sagaLogMock,
+	}
+	state, err := s.StartSaga("testSaga", nil)
+	state, err = s.StartTask(state, "task1")
+	state, err = s.AbortSaga(state)
+	state, err = s.StartCompensatingTask(state, "task1")
+
+	state, err = s.EndCompensatingTask(state, "task1", nil)
+	if err != nil {
+		t.Error(fmt.Sprintf("Expected EndCompensatingTask to not return an error"))
+	}
+}
+
+func TestEndCompTaskLogError(t *testing.T) {
+	entry := EndCompTaskMessageFactory("testSaga", "task1", nil)
+
+	mockCtrl := gomock.NewController(t)
```

Comment:
I disagree, then we are just testing that the fake works correctly (and doing a lot more work writing fakes) inMemorySaga and the demo already does this, but all it proves is that my implementation of inMemorySaga is correct.  Our implementation if DistributedLogSaga could still be wrong.  

Mocks are useful for unit tests to ensure you are passing the right things to the interface, to actually test that the interface implementation is correct & is persisted correctly we need an integration test.